src/ipfs.js

```diff
@@ -14,18 +14,33 @@ module.exports = class Download {
         const ipfs = req.params.ipfs
         const file = req.params[0] ? `${ipfs}${req.params[0]}` : ipfs
         await Blacklist.checkIPFS(ipfs)
-        return res.redirect(`${process.env.S3_URL}/${process.env.S3_BUCKET}/${file}`)
+        return res.redirect(
+          `${process.env.S3_URL}/${process.env.S3_BUCKET}/${file}`
+        )
       } catch (error) {
         next(error)
       }
     }
     this.pin = async (req, res, next) => {
       try {
-        const [x, y, peerId] = [req.params.x, req.params.y, req.params.peerId]
-        if (!isMultihash(peerId)) throw createError(400, `Invalid peerId: ${peerId}`)
+        const [x, y, expectedIPFS, peerId] = [
+          req.params.x,
+          req.params.y,
+          req.params.ipfs,
+          req.params.peerId
+        ]
+        if (!isMultihash(peerId)) {
+          throw createError(400, `Invalid peerId: ${peerId}`)
+        }
         const ipns = await Ethereum.getIPNS(x, y)
         await Download.connectPeer(req.params.peerId)
         const ipfs = await Download.resolveIPNS(ipns)
+        if (ipfs !== expectedIPFS) {
+          throw createError(
+            404,
+            `The IPFS resolved doesn't match with ${expectedIPFS}. Please wait a few minutes and pin again.`
```

Comment:
The message could be improved, may be something like `The resolved IPFS hash doesn't match the expected IPFS hash ${expectedIPFS}. Please wait a few minutes and pin again`